base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -185,36 +188,45 @@ impl MoneroData {
     }
 }
 
-/// Calculate the difficulty attained for the given block deserialized the Monero header from the provided header
-pub fn monero_difficulty(header: &BlockHeader) -> Difficulty {
-    match monero_difficulty_calculation(header) {
-        Ok(v) => v,
-        Err(_) => 0.into(),
-    }
-}
-
 /// Internal function to calculate the difficulty attained for the given block Deserialized the Monero header from the
 /// provided header
-fn monero_difficulty_calculation(header: &BlockHeader) -> Result<Difficulty, MergeMineError> {
+pub fn monero_difficulty(header: &BlockHeader) -> Result<Difficulty, MergeMineError> {
     let monero = MoneroData::new(header)?;
     verify_header(&header, &monero)?;
-    let flags = RandomXFlag::get_recommended_flags();
+
+    debug!(target: LOG_TARGET, "Deserialized Monero data: {:?}", monero);
     let MoneroData {
         key,
         transaction_hashes,
         ..
     } = monero;
 
     let tx_hashes = transaction_hashes.iter().map(Into::into).collect::<Vec<_>>();
-    let input = create_input_blob_from_parts(&monero.header, &tx_hashes)?;
-    let cache = RandomXCache::new(flags, (&key).as_ref())?;
+    let input = &create_input_blob_from_parts(&monero.header, &tx_hashes)?;
+    debug!(target: LOG_TARGET, "RandomX input: {}", input);
+    let input = from_hex(input)?;
+    let key_bytes = from_hex(&key)?;
+    get_random_x_difficulty(&input, &key_bytes).map(|r| r.0)
+}
+
+fn get_random_x_difficulty(input: &[u8], key: &[u8]) -> Result<(Difficulty, Vec<u8>), MergeMineError> {
+    let flags = RandomXFlag::get_recommended_flags();
+    let cache = RandomXCache::new(flags, &key)?;
     let dataset = RandomXDataset::new(flags, &cache, 0)?;
     let vm = RandomXVM::new(flags, Some(&cache), Some(&dataset))?;
-    let hash = vm.calculate_hash((&input).as_ref())?;
-    let scalar = U256::from_big_endian(&hash); // Big endian so the hash has leading zeroes
-    let result = MAX_TARGET / scalar;
+    let hash = vm.calculate_hash(&input)?;
+    debug!(
+        target: LOG_TARGET,
+        "Monero hash: {}, key: {}",
+        hash.to_hex(),
+        Vec::from(key).to_hex()
+    );
+    let scalar = U256::from_little_endian(&hash); // Big endian so the hash has leading zeroes
+    let max_u128 = U256::from(u128::max_value());
+    let max_u256 = max_u128 * max_u128;
+    let result = max_u256 / scalar;
```

Comment:
```suggestion    
    let result = MAX_TARGET / scalar;
```
u128::max_value()*u128::max_value() != U256::Max
Monero uses the eact same value as U256::Max